File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(a, p)
    for _, string in ipairs(a) do
        if p(string) then
            return string:lower()
        end
    end
    return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
    return coroutine.create(function()
        local power = 0
        while true do
            local result = base ^ power
            if result > limit then
                return
            end
            coroutine.yield(result)
            power = power + 1
        end
    end)
end

-- Write your say function here
function say(initial_word)
    local words = {}
    
    local function extra(word)
        if word == nil then
            return table.concat(words, " ")
        end
        table.insert(words, word)
        return extra
    end
    
    return extra(initial_word)
end

-- Write your line count function here

local open = io.open

function meaningful_line_count(path)
    local file = open(path, "r") 
    if not file then 
    error("No such file")
   end
   count = 0
   currentLine = file:read "*l"
   while currentLine ~= nil do
    while string.gsub(currentLine, "%s+", "") == '' or (string.gsub(currentLine, "%s+", "")):sub(1, 1) == "#" do
        currentLine = file:read "*l"
    end
    currentLine = file:read "*l"
    count = count + 1
  end
    
    file:close()
    return count
end

-- Write your Quaternion table here
Quaternion = {}
Quaternion.__index = Quaternion

function Quaternion.new(a, b, c, d)
    return setmetatable({a = a or 0, b = b or 0, c = c or 0, d = d or 0}, Quaternion)
end

function Quaternion:coefficients()
    return {self.a, self.b, self.c, self.d}
end

function Quaternion.__add(q1, q2)
    return Quaternion.new(q1.a + q2.a, q1.b + q2.b, q1.c + q2.c, q1.d + q2.d)
end

function Quaternion.__mul(q1, q2)
    return Quaternion.new(
        q1.a * q2.a - q1.b * q2.b - q1.c * q2.c - q1.d * q2.d,
        q1.a * q2.b + q1.b * q2.a + q1.c * q2.d - q1.d * q2.c,
        q1.a * q2.c - q1.b * q2.d + q1.c * q2.a + q1.d * q2.b,
        q1.a * q2.d + q1.b * q2.c - q1.c * q2.b + q1.d * q2.a
    )
end

function Quaternion:conjugate()
    return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

function Quaternion.__eq(q1, q2)
    return q1.a == q2.a and q1.b == q2.b and q1.c == q2.c and q1.d == q2.d
end

function Quaternion.__tostring(q)
    local function format_term(coef, term)
        if coef == 0 then return "" end
        local str = coef == 1 and "" or coef == -1 and "-" or tostring(coef)
        return (coef > 0 and "+" or "") .. str .. term
    end

    local str = q.a ~= 0 and tostring(q.a) or ""
    str = str .. format_term(q.b, "i")
    str = str .. format_term(q.c, "j")
    str = str .. format_term(q.d, "k")
    
    str = str:gsub("^%+", "")
    return str == "" and "0" or str
end

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(a, p, /):
    for string in a:
        if p(string):
            return string.lower()
    return None


# Write your powers generator here
def powers_generator(*, base, limit):
    power = 0
    while True:
        result = base ** power
        if result > limit:
            break
        yield result
        power += 1


# Write your say function here
def say(word=None, /):
    words = []

    def extra(word=None):
        if word is None:
            return ' '.join(words)
        words.append(word)
        return extra

    return extra(word)

# Write your line count function here
def meaningful_line_count(filename, /):
    try:
        with open(filename, 'r') as file:
            return sum(1 for line in file 
                       if line.strip() and not line.strip().startswith('#'))
    except FileNotFoundError:
        raise

# Write your Quaternion class here
@dataclass(frozen = True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    def __add__(self, other, /):
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)
    
    def __mul__(self, other, /):
        w = self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d
        x = self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c
        y = self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b
        z = self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
        return Quaternion(w, x, y, z)
    
    @property
    def coefficients(self, /):
        return (self.a, self.b, self.c, self.d)
    
    @property
    def conjugate(self, /):
        return Quaternion(self.a, -self.b, -self.c, -self.d)   
    
    def __eq__(self, other, /):
        return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d

    def __str__(self, /):
        components = []

        if self.a != 0.0:
            components.append(f"{self.a}")

        if self.b != 0.0:
            if self.b == 1.0 and not components:
                components.append(f"i")
            elif self.b == 1.0 and components:
                components.append(f"+i")
            elif self.b == -1.0:
                components.append(f"-i")
            elif self.b < 0.0 or not components:
                components.append(f"{self.b}i")
            else:
                components.append(f"+{self.b}i")

        if self.c != 0.0:
            if self.c == 1.0 and not components:
                components.append(f"j")
            elif self.c == 1.0 and components:
                components.append(f"+j")
            elif self.c == -1.0:
                components.append(f"-j")
            elif self.c < 0.0 or not components:
                components.append(f"{self.c}j")
            else:
                components.append(f"+{self.c}j")

        if self.d != 0.0:
            if self.d == 1.0 and not components:
                components.append(f"k")
            elif self.d == 1.0 and components:
                components.append(f"+k")
            elif self.d == -1.0:
                components.append(f"-k")
            elif self.d < 0.0 or not components:
                components.append(f"{self.d}k")
            else:
                components.append(f"+{self.d}k")

        if not components:
            return "0"

        return "".join(components)

        